text/BrunoPitrus/galil.tex

```diff
@@ -0,0 +1,117 @@
+% !TeX spellcheck = pl
+% !TeX program = xelatex
+\documentclass[a4paper]{article}
+\usepackage[margin=12.7mm]{geometry}
+\usepackage{enumitem,polyglossia,amsthm,mathtools,alltt}
+\usepackage[math-style=ISO,colon=literal,mathrm=sym,mathit=sym,mathsf=sym,mathbf=sym,mathtt=sym]{unicode-math}
+\setdefaultlanguage{polish}
+\defaultfontfeatures[\rmfamily,\sffamily,\ttfamily]{}%turn off TeX ligatures
+\setmainfont{Cambria}
+\setmathfont{Cambria Math}
+\setmonofont{Consolas}
+\newtheorem*{theorem}{Twierdzenie}
+\newtheorem*{lemma}{Lemat}
+\setlength{\parindent}{0em}
+\setlength{\parskip}{1em} %Akapity czasem są za krótkie, poniżej 1 linii więc tak będzie lepiej
+
+\begin{document}
+\title{Algorytm Galila–Seiferasa}
+\author{Bruno Pitrus}
+\date{}
+\maketitle
+Algorytm G–S jest algorytmem wyszukiwania wzorca który wymaga jedynie stałej pamięci dodatkowej (eliminując tablicę „mismatch” używaną przez algorytmy typu Morrisa–Pratta) a jednocześnie wciąż działa w asymptotycznym czasie liniowym — osiągając tym samym teoretyczne optimum pod względem zarówno czasu jaki i pamięci.
+
+W tym opisie, tak samo jak w załączonej implementacji algorytmu, stringi indeksowane są od 1. Wszelki kod należy interpretować zgodnie z semantyką języka Python.
+
+Niech \texttt{m = len(word)}, \texttt{n = len(text)}. Rozważmy schemat algorytmu dopasowania wzorca
+\begin{alltt}
+p,q= 0,0
+\textbf{while} True:
+  \textbf{while} text[p+q+1] == word[q+1]: q+= 1
+  \textbf{if} q == m: \textbf{yield} p+1
+  p,q= \textit{p'},\textit{q'}
+\end{alltt}
+W powyższym pseudokodzie \texttt{p} to jest obecny kandydat na pozycję wzorca, a \texttt{q} to prefiks o którym wiemy że się zgadza.
+
+Naiwny algorytm \textit{brute-force} używa \texttt{p'= p+1}, \texttt{q'= 0}. Ponieważ wiemy jednak że \texttt{text[p+1:p+q+1] == word[1:q+1]} to rozważanie \texttt{p'= p+x} ma sens tylko jeśli \texttt{word[1:q-x+1] == word[x+1:q+1]}; na tej obserwacji jest oparty algorytm Knutha–Morrisa–Pratta, obliczający \texttt{p'= p+shift[q], q'= q-shift[q] if q>0 else 0}, gdzie \texttt{shift[q]} jest długością najkrótszego okresu \texttt{word[1:q+1]}.
+
+Algorytm Galila–Seiferasa jest oparty na tym schemacie oraz następującym twierdzeniu:
+\begin{theorem}[O dekompozycji]
+Dla ustalonego (odpowiednio dużego) \(k\), każde słowo \texttt{x} można przedstawić jako \texttt{u+v}, gdzie \texttt{v} ma co najwyżej jeden \textit{\(k\)-okres prefiksu} a \texttt{u} jest długości \(\symrm O{\left(\text{długość najkrótszego okresu \texttt{v}}\right)}\). 

```

Comment:
Two questions:
1. Is there any precise bound in the theorem instead of O(period(v))?
2. Is the proof long and complicated enough not to put there at least a gist of it?